# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/public/CompileOptions.h
# Commit: 4107b758e7ae
# Full Hash: 4107b758e7aea25d4529511f47bbd54531640dc4
# Author: Denis Palmeiro <dpalmeiro@mozilla.com>
# Date: 2020-05-01 09:42:47
# Regressor Bug: 1501608
# File Overlap Count: 1
# Description:
#   Bug 1501608 - Remove the ELEMENT_SLOT in the ScriptSourceObject and instead use a callback function to return the script element based on the value of the privateValue in the SSO. r=jonco,smaug
#   
#   We can reduce the size of the SSO by removing the element slot entirely, and instead retrieve the element through a callback function.  The callback will take in the value in the private slot of the SSO, which is either a LoadedScript* (from the browser) or a JSObject* (from the shell).  In addition, this removes the requirement of having a script dom element ready when parsing a JS script which can open up new opportunities for performance.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D70417
# ==============================================================================

diff -r d3a637853a36 -r 4107b758e7ae js/public/CompileOptions.h
--- a/js/public/CompileOptions.h	Thu Apr 30 14:32:56 2020 +0000
+++ b/js/public/CompileOptions.h	Thu Apr 30 16:21:23 2020 +0000
@@ -61,6 +61,7 @@
 #include "jstypes.h"  // JS_PUBLIC_API
 
 #include "js/RootingAPI.h"  // JS::PersistentRooted, JS::Rooted
+#include "js/Value.h"
 
 struct JS_PUBLIC_API JSContext;
 class JS_PUBLIC_API JSObject;
@@ -151,7 +152,7 @@
   const char* filename() const { return filename_; }
   const char* introducerFilename() const { return introducerFilename_; }
   const char16_t* sourceMapURL() const { return sourceMapURL_; }
-  virtual JSObject* element() const = 0;
+  virtual Value privateValue() const = 0;
   virtual JSString* elementAttributeName() const = 0;
   virtual JSScript* introductionScript() const = 0;
 
@@ -241,17 +242,17 @@
  * anything else it entrains, will never be freed.
  */
 class JS_PUBLIC_API OwningCompileOptions final : public ReadOnlyCompileOptions {
-  PersistentRooted<JSObject*> elementRoot;
   PersistentRooted<JSString*> elementAttributeNameRoot;
   PersistentRooted<JSScript*> introductionScriptRoot;
   PersistentRooted<JSScript*> scriptOrModuleRoot;
+  PersistentRooted<Value> privateValueRoot;
 
  public:
   // A minimal constructor, for use with OwningCompileOptions::copy.
   explicit OwningCompileOptions(JSContext* cx);
   ~OwningCompileOptions();
 
-  JSObject* element() const override { return elementRoot; }
+  Value privateValue() const override { return privateValueRoot; }
   JSString* elementAttributeName() const override {
     return elementAttributeNameRoot;
   }
@@ -281,10 +282,10 @@
 class MOZ_STACK_CLASS JS_PUBLIC_API CompileOptions final
     : public ReadOnlyCompileOptions {
  private:
-  Rooted<JSObject*> elementRoot;
   Rooted<JSString*> elementAttributeNameRoot;
   Rooted<JSScript*> introductionScriptRoot;
   Rooted<JSScript*> scriptOrModuleRoot;
+  Rooted<Value> privateValueRoot;
 
  public:
   // Default options determined using the JSContext.
@@ -293,16 +294,16 @@
   // Copy the transitive options from another options object.
   CompileOptions(JSContext* cx, const TransitiveCompileOptions& rhs)
       : ReadOnlyCompileOptions(),
-        elementRoot(cx),
         elementAttributeNameRoot(cx),
         introductionScriptRoot(cx),
-        scriptOrModuleRoot(cx) {
+        scriptOrModuleRoot(cx),
+        privateValueRoot(cx) {
     copyPODTransitiveOptions(rhs);
 
     filename_ = rhs.filename();
     introducerFilename_ = rhs.introducerFilename();
     sourceMapURL_ = rhs.sourceMapURL();
-    elementRoot = rhs.element();
+    privateValueRoot = rhs.privateValue();
     elementAttributeNameRoot = rhs.elementAttributeName();
     introductionScriptRoot = rhs.introductionScript();
     scriptOrModuleRoot = rhs.scriptOrModule();
@@ -315,7 +316,7 @@
     copyPODNonTransitiveOptions(rhs);
   }
 
-  JSObject* element() const override { return elementRoot; }
+  Value privateValue() const override { return privateValueRoot; }
 
   JSString* elementAttributeName() const override {
     return elementAttributeNameRoot;
@@ -348,8 +349,8 @@
     return *this;
   }
 
-  CompileOptions& setElement(JSObject* e) {
-    elementRoot = e;
+  CompileOptions& setPrivateValue(const Value& v) {
+    privateValueRoot = v;
     return *this;
   }
 